### DISCUSSION TITLE
Multi-edge graph and pre_transofrms=T.KNNGraph

### DISCUSSION THREAD

**Discussion by ghost** [pyg-team/pytorch_geometric#5895] [Q&A]
Hi PyG Team,

I am doing an experiment where I have created edge_index from a multiedge networkx graph and then I pre-transformed using KNN graph with k=3. The results improved and especially the roc_auc curve. I am trying to understand what impact this has on the input graph? Does pre-transforming to KNN change multi edge directed graph to undirected/ directed graph, or is it that the multi graph remains the same? I have edge features and doing edge classification. I have noticed that the number of edges after using T.KNNGraph pretransform remian the same as was the multiedge graph construction. I have following queries that I have not been able to understand and would appreciate your help:

1) Since k=3 means that every node will retain edges/connections with 3 nearest neighbours, does that mean that this setting will lose connection and in turn edge features if neighbours otherwise were more than 3?

I am doing edge classification , so retaining edges and edge features is an important task. My concept of applying KNN was to include more information related to graph construction and relations in the graph learning.

TIA
Tanzi

**Comment by rusty1s** [comment]
If I understand you correctly, you have an initial graph with edge features. Instead of using these, you obtain a k-NN graph from node features, and using this as GNN input. Is that correct?

It looks like you don't use any of the edge features this way (potentially you do in the final edge classifier), but the k-NN will basically throw away all information about previous edges. If you want to maintain original edges, I suggest to fuse both graph connections together. WDYT?

**Comment by ghost** [comment]
I’m not using any node features, all the learning is suppose to be happen through edge features. I got good results using multi edge graph structure at input. Pre-transforming to KNN graph improves the results slightly better. I researched about knn graphs and could not find a lot in terms of edge data. I’m unsure what’s happening.

**Reply by rusty1s** [reply]
How do you maintain edge features when using KNN graph? Not sure I understand, sorry! Can you give some insights into your model architecture?

**Comment by ghost** [comment]
To start with, I created PyG data object feeding in edge_index, edge features as edge attributes, node features which is a constant matrix, and Knn graph pre_transform. Probably, the knn transform is not doing much here. .. ??

**Reply by rusty1s** [reply]
If all your node features are constant, then the knn will just randomly connect to some other nodes. Yes, it is not doing much in this case :(
